File: php/src/Resource/Workers.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Model\PatchWorker;
use Fivexer\SDK\Model\UpsertWorker;
use Fivexer\SDK\Model\WorkerAvailability;
use Fivexer\SDK\Model\WorkerDetail;
use Fivexer\SDK\Model\WorkerList;
use Fivexer\SDK\Model\WorkerMetrics;
use Fivexer\SDK\Model\WorkerQueue;
use Fivexer\SDK\Model\WorkerTimeEntriesResult;

/**
 * Worker management resource. Delegates to Fivexer::request() so all transport
 * logic (auth, retry, quota, errors) lives in one place.
 */
final class Workers
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    /** Create or update a worker; returns the worker id (server-assigned if absent). */
    public function upsert(?UpsertWorker $worker = null): string
    {
        $data = $this->client->request('POST', '/workers', ($worker ?? new UpsertWorker())->toArray()) ?? [];
        return (string) $data['id'];
    }

    /** List all worker ids in the workspace. */
    public function list(): WorkerList
    {
        return WorkerList::fromArray($this->client->request('GET', '/workers') ?? []);
    }

    /** A worker's current task queue (tentative matches). */
    public function queue(string $workerId): WorkerQueue
    {
        return WorkerQueue::fromArray(
            $this->client->request('GET', '/workers/' . \rawurlencode($workerId) . '/queue') ?? [],
        );
    }

    /**
     * The operator-plane mirror of the worker's own portal metrics: today plus a rolling window.
     * Worked time comes from the shift log — shift time minus overlapping breaks — so both planes
     * report the same figure. Requires the control plane.
     *
     * @param string|null $window rolling window such as '7d' (the default) up to '30d', or null
     */
    public function metrics(string $workerId, ?string $window = null): WorkerMetrics
    {
        return WorkerMetrics::fromArray(
            $this->client->request(
                'GET',
                '/workers/' . \rawurlencode($workerId) . '/metrics',
                null,
                ['window' => $window],
            ) ?? []
        );
    }

    /**
     * The recorded shift and break log for a window (default: the last 7 days) — the working-time
     * record an EU employer must keep (CJEU C-55/18). A shift is a recorded stretch of
     * availability and breaks are time inside one, so `totals->workingMs` is `onShiftMs` minus
     * `breakMs`. Requires the control plane.
     *
     * @param string|null $from ISO-8601 start, or null
     * @param string|null $to ISO-8601 end, or null
     */
    public function timeEntries(
        string $workerId,
        ?string $from = null,
        ?string $to = null,
    ): WorkerTimeEntriesResult {
        return WorkerTimeEntriesResult::fromArray(
            $this->client->request(
                'GET',
                '/workers/' . \rawurlencode($workerId) . '/time-entries',
                null,
                ['from' => $from, 'to' => $to],
            ) ?? []
        );
    }

    /** Remove a worker and clear their backlog. */
    public function remove(string $workerId): void
    {
        $this->client->request('DELETE', '/workers/' . \rawurlencode($workerId));
    }

    /** Full detail of one worker, including their current load against their cap. */
    public function get(string $workerId): WorkerDetail
    {
        return WorkerDetail::fromArray(
            $this->client->request('GET', '/workers/' . \rawurlencode($workerId)) ?? []
        );
    }

    /** Partial update — fields left unset keep their stored value. Returns the worker id. */
    public function patch(string $workerId, PatchWorker $patch): string
    {
        $data = $this->client->request('PATCH', '/workers/' . \rawurlencode($workerId), $patch->toArray()) ?? [];
        return (string) $data['id'];
    }

    /**
     * Pause or resume a worker. Pausing preserves the unaccepted backlog ("back in ten
     * minutes").
     *
     * @param bool $releaseBacklog When pausing, also requeue the unaccepted backlog so others
     *     inherit it now ("gone for the day"). Accepted work in progress is never touched.
     *     Only valid when pausing — the API rejects it on a resume, so it is sent only when true.
     */
    public function setAvailability(
        string $workerId,
        bool $available,
        bool $releaseBacklog = false,
    ): WorkerAvailability {
        $body = ['available' => $available];
        if ($releaseBacklog) {
            $body['releaseBacklog'] = true;
        }
        return WorkerAvailability::fromArray(
            $this->client->request('POST', '/workers/' . \rawurlencode($workerId) . '/availability', $body) ?? []
        );
    }
}

File: java/src/main/java/io/fivexer/sdk/Workers.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.PatchWorker;
import io.fivexer.sdk.model.UpsertWorker;
import io.fivexer.sdk.model.WorkerAvailability;
import io.fivexer.sdk.model.WorkerDetail;
import io.fivexer.sdk.model.WorkerList;
import io.fivexer.sdk.model.WorkerMetrics;
import io.fivexer.sdk.model.WorkerQueue;
import io.fivexer.sdk.model.WorkerRef;
import io.fivexer.sdk.model.WorkerTimeEntriesResult;

/** Worker registration, configuration and availability. */
public final class Workers {

    private final Fivexer client;

    Workers(Fivexer client) {
        this.client = client;
    }

    /** Create or replace a worker. Returns the id, which the API generates when not supplied. */
    public String upsert(UpsertWorker worker) {
        UpsertWorker input = worker == null ? new UpsertWorker() : worker;
        WorkerRef ref = client.request("POST", "/workers", input.toJson(), null, WorkerRef.class, false);
        return ref.getId();
    }

    public WorkerList list() {
        return client.request("GET", "/workers", null, null, WorkerList.class, false);
    }

    public WorkerDetail get(String workerId) {
        return client.request("GET", "/workers/" + Json.enc(workerId), null, null, WorkerDetail.class, false);
    }

    /** Partial update — fields left unset keep their stored value. */
    public String patch(String workerId, PatchWorker patch) {
        WorkerRef ref = client.request("PATCH", "/workers/" + Json.enc(workerId), patch.toJson(), null,
                WorkerRef.class, false);
        return ref.getId();
    }

    /**
     * Pause or resume a worker. Pausing preserves the unaccepted backlog ("back in ten minutes").
     */
    public WorkerAvailability setAvailability(String workerId, boolean available) {
        return setAvailability(workerId, available, false);
    }

    /**
     * @param releaseBacklog when pausing, also requeue the unaccepted backlog so others inherit
     *     it now ("gone for the day"). Accepted work in progress is never touched. Only valid
     *     when pausing — the API rejects it on a resume, so it is sent only when true.
     */
    public WorkerAvailability setAvailability(String workerId, boolean available, boolean releaseBacklog) {
        com.google.gson.JsonObject body = new com.google.gson.JsonObject();
        body.addProperty("available", available);
        if (releaseBacklog) {
            body.addProperty("releaseBacklog", true);
        }
        return client.request("POST", "/workers/" + Json.enc(workerId) + "/availability", body.toString(),
                null, WorkerAvailability.class, false);
    }

    public WorkerQueue queue(String workerId) {
        return client.request("GET", "/workers/" + Json.enc(workerId) + "/queue", null, null,
                WorkerQueue.class, false);
    }

    /** Recorded time and throughput for one worker over the default window (7 days). */
    public WorkerMetrics metrics(String workerId) {
        return metrics(workerId, null);
    }

    /**
     * The operator-plane mirror of the worker's own portal metrics: today plus a rolling window.
     * Worked time comes from the shift log — shift time minus overlapping breaks — so both planes
     * report the same figure. Requires the control plane.
     *
     * @param window rolling window such as {@code "7d"} (the default) up to {@code "30d"}, or null
     */
    public WorkerMetrics metrics(String workerId, String window) {
        return client.request("GET", "/workers/" + Json.enc(workerId) + "/metrics", null,
                Json.query("window", window), WorkerMetrics.class, false);
    }

    /** The recorded shift and break log for the default window (the last 7 days). */
    public WorkerTimeEntriesResult timeEntries(String workerId) {
        return timeEntries(workerId, null, null);
    }

    /**
     * The recorded shift and break log for a window — the working-time record an EU employer must
     * keep (CJEU C-55/18). A shift is a recorded stretch of availability and breaks are time
     * inside one, so {@code totals.workingMs} is {@code onShiftMs} minus {@code breakMs}.
     * Requires the control plane.
     *
     * @param from ISO-8601 start, or null; @param to ISO-8601 end, or null
     */
    public WorkerTimeEntriesResult timeEntries(String workerId, String from, String to) {
        return client.request("GET", "/workers/" + Json.enc(workerId) + "/time-entries", null,
                Json.query("from", from, "to", to), WorkerTimeEntriesResult.class, false);
    }

    public void remove(String workerId) {
        client.request("DELETE", "/workers/" + Json.enc(workerId), null, null, Void.class, true);
    }
}
